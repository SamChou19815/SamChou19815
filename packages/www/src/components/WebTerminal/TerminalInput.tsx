import React, { ReactElement, KeyboardEvent, RefObject, useState } from 'react';
import autoComplete from './auto-complete';
import styles from './Terminal.module.css';

type Props = {
  readonly terminalInput: RefObject<HTMLInputElement>;
  readonly onArrow: (arrow: 'up' | 'down') => string | null;
  readonly onSubmit: (line: string) => void;
};

export default ({ terminalInput, onArrow, onSubmit }: Props): ReactElement => {
  const [line, setLine] = useState('');

  const onKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    switch (event.key) {
      case 'Enter':
        setLine('');
        onSubmit(line);
        break;
      case 'ArrowUp':
      case 'ArrowDown': {
        const result = onArrow(event.key === 'ArrowUp' ? 'up' : 'down');
        if (result !== null) {
          setLine(result);
        }
        break;
      }
      case 'Tab':
        event.preventDefault();
        setLine(currentLine => autoComplete(currentLine));
        break;
      default:
        break;
    }
  };

  return (
    <div className={styles.TerminalInputArea}>
      <span className={styles.TerminalPromptLabel}>$</span>
      <input
        name="terminal-input"
        ref={terminalInput}
        className={styles.TerminalInput}
        value={line}
        onChange={event => setLine(event.currentTarget.value)}
        onKeyDown={onKeyDown}
        type="text"
        autoComplete="off"
      />
    </div>
  );
};
