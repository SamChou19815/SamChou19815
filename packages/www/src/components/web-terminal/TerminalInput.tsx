import React, { KeyboardEvent, RefObject, useState, useEffect } from 'react';

import { useWebTerminalCommands } from './WebTerminalCommandsContext';
import autoComplete from './auto-complete';

type Props = {
  readonly terminalInput: RefObject<HTMLInputElement>;
  readonly onArrow: (arrow: 'up' | 'down') => string | null;
  readonly onSubmit: (line: string) => void;
};

const TerminalInput = ({ terminalInput, onArrow, onSubmit }: Props): JSX.Element => {
  const commands = useWebTerminalCommands();
  const [line, setLine] = useState('');
  const [justProcessedArrowKey, setJustProcessedArrowKey] = useState(false);

  const onKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    switch (event.key) {
      case 'Enter':
        setLine('');
        onSubmit(line);
        break;
      case 'ArrowUp':
      case 'ArrowDown': {
        const result = onArrow(event.key === 'ArrowUp' ? 'up' : 'down');
        setJustProcessedArrowKey(true);
        if (result !== null) {
          setLine(result);
        }
        break;
      }
      case 'Tab':
        event.preventDefault();
        setLine((currentLine) => autoComplete(commands, currentLine));
        break;
      default:
        break;
    }
  };

  useEffect(() => {
    const terminalInputDOMNode = terminalInput?.current;
    if (terminalInputDOMNode != null && justProcessedArrowKey) {
      setJustProcessedArrowKey(false);
      terminalInputDOMNode.selectionStart = line.length;
    }
  }, [terminalInput, line, justProcessedArrowKey]);

  return (
    <div className="inline-flex justify-center w-full">
      <span className="text-blue-500 pr-2">$</span>
      <input
        name="terminal-input"
        ref={terminalInput}
        className="border-0 flex-grow w-full h-6 bg-transparent font-mono outline-none"
        value={line}
        onChange={(event) => setLine(event.currentTarget.value)}
        onKeyDown={onKeyDown}
        type="text"
        autoComplete="off"
      />
    </div>
  );
};

export default TerminalInput;
