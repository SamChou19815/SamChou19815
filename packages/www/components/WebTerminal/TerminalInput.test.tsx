import React, { ReactElement, useRef } from 'react';
import ReactDOM from 'react-dom';

import { act } from 'react-dom/test-utils';

import TerminalInput from './TerminalInput';

it('TerminalInput can respond to inputs.', () => {
  const container = document.createElement('div');
  document.body.appendChild(container);

  let upTyped = false;
  let downTyped = false;
  let enterTyped = false;

  const onArrow = (direction: 'up' | 'down'): string | null => {
    if (direction === 'up') {
      upTyped = true;
      return null;
    }
    downTyped = true;
    return 'ah!';
  };
  const onSubmit = (): void => {
    enterTyped = true;
  };

  const Wrapper = (): ReactElement => {
    const terminalInput = useRef<HTMLInputElement>(null);
    return <TerminalInput terminalInput={terminalInput} onArrow={onArrow} onSubmit={onSubmit} />;
  };

  act(() => {
    ReactDOM.render(<Wrapper />, container);
  });
  const inputNode = document.querySelectorAll('[name="terminal-input"]')[0] as HTMLInputElement;

  act(() => {
    inputNode.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true }));
  });
  expect(enterTyped).toBeTruthy();

  act(() => {
    inputNode.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp', bubbles: true }));
  });
  expect(upTyped).toBeTruthy();

  act(() => {
    inputNode.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown', bubbles: true }));
  });
  expect(downTyped).toBeTruthy();
  expect(inputNode.value).toBe('ah!');

  act(() => {
    inputNode.dispatchEvent(new KeyboardEvent('keydown', { key: 'Tab', bubbles: true }));
  });
  expect(inputNode.value).toBe('ah!');

  act(() => {
    inputNode.dispatchEvent(new KeyboardEvent('keydown', { key: 'Shift', bubbles: true }));
  });

  document.body.removeChild(container);
});
