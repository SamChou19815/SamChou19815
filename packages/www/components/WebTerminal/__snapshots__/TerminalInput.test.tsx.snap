// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TerminalInput matches snapshot. 1`] = `
<div
  className="TerminalInputArea"
>
  <span
    className="TerminalPromptLabel"
  >
    $
  </span>
  <input
    autoComplete="off"
    className="TerminalInput"
    name="terminal-input"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
</div>
`;
